contributors/design-proposals/storage/local-volume-management.md

```diff
@@ -0,0 +1,219 @@
+# Local volume management
+
+**Authors:** [mlmhl](https://github.com/mlmhl)
+
+- [Background](#background)
+- [Object](#object)
+- [Roadmap](#roadmap)
+- [Implementation Design](#design)
+	- [Kubernetes Side Refactor](#k8s_refactor)
+		- [Volum Plugin Interfaces Refactor](#k8s_interfaces_refactor)
+		- [Kubelet Volume Manager Refactor](#k8s_vm_refactor)
+			- [Mount Device Operation](#k8s_vm_mount)
+			- [Umount Device Operation](#k8s_vm_umount)
+		- [A/D Controller Refactor](#k8s_ad_controller_refactor)
+	- [Volume Plugin Side Refactor](#volume_plugin_refactor)
+		- [Mount Device Operation](#volume_plugin_mount_refactor)
+		- [Umount Device Operation](#volume_plugin_umount_refactor)
+
+## <a name="background"></a>Background
+
+For all attachable volume plugins, there are three major steps when provisioning a volume for a `Pod`:
+
+1. Attach the volume to according node the pod is scheduled to
+2. Mount the attached device to a global mount path
+3. Bind the global mount path to a pod specified path
+
+The recycle of a volume is similar, umount/detach the path/device in a reverse order.
+
+Today, A/D controller is responsible for the attach/detach operation of volumes, kubelet is responsible for the other two steps. Before we start, let's make a quick review of the whole workflow:
+
+1. A/D controller watches all scheduled `Pods` from apiserver, and add all volumes in `Pod.Spec.Volumes` to its `DesiredStateOfWorld`. This cache maintains volumes need to be attached
+2. A/D controller periodically checks volumes in `DesiredStateOfWorld` to make sure if they are attached. If not, controller will invoke the attach operation of this volume. After attach operation finished, controller will add this volume to `Node.Status.VolumesAttached`
+3. Kubelet watches all pod scheduled to its own, and add all volumes in `Pod.Spec.Volumes` to its `DesiredStateOfWorld`
+4. Kubelet periodically checks volumes in `DesiredStateOfWorld` to make sure if they are mounted. If not, kubelet will:
+	- Wait for the attach operation finished
+	- Mount the attached device to the global mount path
+	- Bind the global mount path to a pod specified path
+
+This workflow if nicely for all cloud volumes(`AWS EBS`, `GCE PD`, etc.) which have a remote attach API, but not very suitable for local volumes(`fc`, `iscsi`, `rbd`, etc.) which need a local attach operation. So current implementation make the `Attach/Detach` method of these local volumes do nothing, and execute the real attach operations in `WaitForAttach` method. This led to the following problems:
```

Comment:
If i understand your proposal correctly, you just extract the `MountDevice` out of `Attacher` and then put it into `DeviceMounter`.  Can we really solve this problem ?  We still need to execute `attach` for rbd in `WaitForAttach`.

Reply:
oh, you mentioned at the end of this proposal, move `attach` to `MountDevice` ?